source/aws_http_credentials_provider.c

```diff
@@ -0,0 +1,612 @@
+    if (aws_tls_connection_options_set_server_name(&tls_connection_options, allocator, &host)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): failed to create a tls connection options with error %s",
+            (void *)provider,
+            aws_error_str(aws_last_error()));
+        goto on_error;
+    }
+
+    struct aws_socket_options socket_options;
+    AWS_ZERO_STRUCT(socket_options);
+    socket_options.type = AWS_SOCKET_STREAM;
+    socket_options.domain = AWS_SOCKET_IPV4;
+    socket_options.connect_timeout_ms = (uint32_t)aws_timestamp_convert(
+        HTTP_CONNECT_TIMEOUT_DEFAULT_IN_SECONDS, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_MILLIS, NULL);
+
+    struct aws_http_connection_manager_options manager_options;
+    AWS_ZERO_STRUCT(manager_options);
+    manager_options.bootstrap = options->bootstrap;
+    manager_options.initial_window_size = HTTP_RESPONSE_SIZE_LIMIT;
+    manager_options.socket_options = &socket_options;
+    manager_options.host = host;
+    manager_options.port = 443;
+    manager_options.max_connections = 2;
+    manager_options.shutdown_complete_callback = s_on_connection_manager_shutdown;
+    manager_options.shutdown_complete_user_data = provider;
+    manager_options.tls_connection_options = &tls_connection_options;
+
+    impl->function_table = options->function_table;
+    if (impl->function_table == NULL) {
+        impl->function_table = g_aws_credentials_provider_http_function_table;
+    }
+
+    impl->token_provider = aws_credentials_provider_acquire(options->token_provider);
+    impl->endpoint = aws_string_new_from_string(allocator, options->endpoint);
+    impl->account_id = aws_string_new_from_string(allocator, options->account_id);
+    impl->connection_manager = impl->function_table->aws_http_connection_manager_new(allocator, &manager_options);
+    if (impl->connection_manager == NULL) {
+        goto on_error;
+    }
+
+    provider->shutdown_options = options->shutdown_options;
+
+    if (options->retry_strategy == NULL) {
+        struct aws_standard_retry_options retry_options = {
+            .backoff_retry_options =
+                {
+                    .el_group = options->bootstrap->event_loop_group,
+                    .max_retries = HTTP_MAX_ATTEMPTS,
+                },
+        };
+
+        impl->retry_strategy = aws_retry_strategy_new_standard(allocator, &retry_options);
+    } else {
+        impl->retry_strategy = options->retry_strategy;
+    }
+
+    if (!impl->retry_strategy) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): failed to create a retry strategy with error %s",
+            (void *)provider,
+            aws_error_debug_str(aws_last_error()));
+        goto on_error;
+    }
+
+    aws_tls_connection_options_clean_up(&tls_connection_options);
+    return AWS_OP_SUCCESS;
+on_error:
+    aws_credentials_provider_destroy(provider);
```

Comment:
hmmm, so when the initialization failed, we will destroy the provider here, which will actually free the memory for `provider` as well, according to https://github.com/awslabs/aws-c-auth/pull/276/files#diff-7df01304b24de64755d81ee19179ffcbb204080bfa33965734cf521989c455a5R476.

This seems like a concern, and may cause double free from https://github.com/awslabs/aws-c-auth/pull/276/files#diff-23acef6a19dca102219feb72adf38a73142c400f8870d91dfcde1f6a7a43cf64R286?

I'd suggest we only clean up the impl here, and leave the passed-in `provider` as it is, so that it's caller's responsibility to properly handle the clean up of the passed-in arg.